### ISSUE TITLE
NumberFilter raise error when I trying to send NaN as value

### ISSUE THREAD

**Issue by yakimka** [carltongibson/django-filter#1399]
I did make an example project. You can see it here https://github.com/yakimka/nan_example
I use django 3.2.5, django-filter 2.4.0 and djangorestframework 3.12.4

## Steps to reproduce:

1. Clone repository
2. Install requirements: `pip install -r requirements.txt`
3. Make migrations: `python manage.py migrate`
4. Run server: `python manage.py runserver 127.0.0.1:8080`
5. Make request to order list endpoint with sum=NaN parameter (with Postman or curl `curl --location --request GET 'http://127.0.0.1:8080/orders/?sum=NaN'`)
6. See exception

![exception](https://user-images.githubusercontent.com/28621349/125784419-ca624278-ce47-492f-8c73-60afd5e21f3c.png)

## Expected result:

Validation error as if I was sending a non-number

```json
{
    "sum": [
        "Enter a number."
    ]
}
```

**Comment by carltongibson** [comment]
Ok, sounds valid at first glance. 

Next steps: Can you look at the generated form here? What's the field? How does that handle NaN?

**Comment by yakimka** [comment]
NumberFilter uses forms.DecimalField, NaN for decimal is a valid value, but for NumberFilter "Not a Number" is not valid, I think. So maybe we need to add a validator to NumberFilter that can handle NaN value.

**Comment by carltongibson** [comment]
TBH I think we should have a validator on the form to reject this case. 🤔

**Comment by yakimka** [comment]
I updated the example project.
It also raises an exception with forms.

![form_exception](https://user-images.githubusercontent.com/28621349/125800071-597b7dde-ae49-46a7-b0ac-0bab3533e23a.png)

Form rendered with attribute type="number", but I can delete this attribute, or just use query parameter

**Comment by carltongibson** [comment]
What's the form look like there?

**Comment by yakimka** [comment]
Form:
![image](https://user-images.githubusercontent.com/28621349/125802124-ea447ced-9f16-4ed1-bb39-71bd179520f4.png)

Code:
```
<!doctype html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport"
          content="width=device-width, user-scalable=no, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
</head>
<body>
    <form method="get">
        {{ filter.form.as_p }}
        <input type="submit" />
    </form>
    {% for obj in filter.qs %}
        {{ obj.name }} - ${{ obj.price }}<br />
    {% endfor %}
</body>
</html>
```

**Comment by carltongibson** [comment]
Sorry, I meant the Django form on the FilterSet... I'll need to have a look at the project but we should be able to add a Django validators to that to reject the NaN value.

**Comment by yakimka** [comment]
I think I found the source of the error - this is Django 😃 

`forms.DecimalField` use `django.core.validators.DecimalValidator` by default. This validator raises `ValidationError("Enter a number.")` on "inf" and "NaN" values. But in the `run_validators` method of the `Field` class Django calls all validators, even if some validator already caused an `ValidationError` https://github.com/django/django/blob/0eca7a66239ef646f59fe2af643199275dae7a35/django/forms/fields.py#L140

So, when we run NumberFilter validation it will raise validation error in DecimalValidator, and then fail in `MaxValueValidator` because Decimal("NaN") can't be compared with numbers.

We can reproduce this behavior with Django forms itself:

```python
class OrderForm(forms.ModelForm):
    sum = DecimalField(max_value=12)

    class Meta:
        model = Order
        fields = ['sum']
```

This form also will raise `InvalidOperationError` if we pass "NaN" string as value

**Comment by yakimka** [comment]
I made PR to Django https://github.com/django/django/pull/14645

**Comment by carltongibson** [comment]
Upstream ticket https://code.djangoproject.com/ticket/32949 — Regression in Django 3.2. Should be fixed in 3.2.6. 
Thanks @yakimka .
